### PR TITLE
MRG: add --module-name argument to cython command

### PR DESCRIPTION
**Pull request by matthew-brett**
I'd like to ask for feedback for this implementation of a
`--module-name` command line argument to the `cython` utility.

It can be useful to specify the module name for the output file
directly, rather than working it out from the enclosing file tree -
particularly for out of tree build systems, like Meson.

See: https://github.com/rgommers/scipy/issues/31#issuecomment-1002662816
for background.

However, the implementation here is a little ugly, because the cython
command line utility always executes through the `compile_multiple` code
path in `Main.py`.  This in turn is because the passed `sources`
argument is a list, and not a string, and therefore, the check around
line 629 always fails, meaning that I have to recover the passed
`module_name` from `options` in `compile_multiple`, rather than using
the `full_module_name` argument available in `compile_single`.

Is this the right thing to do?  If so, I'll write up some tests.

### CODE REVIEW

**Comment by rgommers** [comment]
Nice, thanks @matthew-brett 

> This in turn is because the passed `sources`
> argument is a list, and not a string, and therefore, the check around
> line 629 always fails, meaning that I have to recover the passed
> `module_name` from `options` in `compile_multiple`, rather than using
> the `full_module_name` argument available in `compile_single`.

I noticed this too, and had assumed that was a small implementation issue somewhere, that hasn't been noticed before because taking the `compile_multiple` path works just fine. The alternative fix is to unpack `sources` if it's a length-1 list (or don't turn the input into a list in the first place) - that would perhaps be cleaner, however this diff is only a few lines long and should have no side-effects so is perhaps fine and safer.

**Comment by matthew-brett** [comment]
Reviewing the code for `compile_single` and `compile_multiple`, I believe it would be equivalent to use `compile_single` for all calls to the `cython` utility where there was only one passed source file - and this makes the change neater, because I can use the`full_module_name` argument directly.   On the other hand, it does mean that I can't use the `--module-name` with `--timestamps`, as written - and if I'm wrong about the equivalence, it will change the behavior of `cython` calls for single sources.

**Comment by matthew-brett** [comment]
On further investigation, there is a difference in behavior in the code path for `compile_single` and `compile_multi`.  I don't know whether it is intentional, but `compile_multi` first converts all source files to absolute paths, and this in turn will make a difference in the found module name.  Both call into `run_pipeline`, which then, in turn, call `extract_module_name`, and the results differ for relative and absolute path names.  For example, running the code from a submodule folder (in this case `nibabel/cifti2`, where both `nibabel` and `nibabel/cifti2` contain `__init__.py` files):

```
[ins] In [14]: context.extract_module_name('foo.pyx', None)
Out[14]: 'foo'

[ins] In [15]: context.extract_module_name(os.path.abspath('foo.pyx'), None)
Out[15]: 'nibabel.cifti2.foo'
```

I think this is accidental, and there's a bug in the `extract_module_name` code, but nevertheless, as things stand, `compile_single` does not exactly correspond to `compile_multi` with a single filename, and so we need the approach in this PR to avoid changes in the default behavior of `cython`.

**Comment by matthew-brett** [comment]
I have added some tests, and some fixes - I believe it is ready for review.

**Comment by matthew-brett** [comment]
Tests passing.

**Comment by rgommers** [comment]
This LGTM (the feature, I'm not familiar with this way of testing). 

@matthew-brett did you already try use this with SciPy? There's a bunch of direct `cython` invocations where you could add this argument (search for `fast-fail` to find them). It should allow getting rid of hacks like this:
```meson
# Needed to trick Cython, it won't do a relative import outside a package
_dummy_init_optimize = custom_target('_dummy_init_optimize',
  output: '__init__.py',
  input: '__init__.py',
  command: [copier, '@INPUT@', '@OUTDIR@']
)
```

**Comment by matthew-brett** [comment]
This does work as intended, in testing with Scipy - discussions over at https://github.com/rgommers/scipy/issues/31

File: Cython/Compiler/Tests/TestCmdLine.py [hunk 1]
```diff
@@ -505,7 +520,6 @@ def error(*args):
                 sys.stderr = old_stderr
             self.assertTrue(stderr.getvalue())
 
-        error('-1')
```
Comment (scoder) [inline Cython/Compiler/Tests/TestCmdLine.py:508]: Why was this removed?

File: Cython/Compiler/Tests/TestCmdLine.py [hunk 2]
```diff
@@ -514,3 +528,9 @@ def error(*args):
         error('--verbose=1')
         error('--cleanup')
         error('--debug-disposal-code-wrong-name', 'file3.pyx')
+        # No source file (source file appears to be module name).
+        error('--module-name', 'foo.pyx')
+        # Cannot use --module-name with more than one source file.
+        error('--module-name', 'foo.bar', 'foo.pyx', 'bar.pyx')
+        # Cannot use --module-name with --timestamps
+        error('--module-name', 'foo.bar', '--timestamps', 'foo.pyx')
```
Comment (scoder) [inline Cython/Compiler/Tests/TestCmdLine.py:536]: Would be nice to start (regex) testing for error messages here. Better than adding them in comments.

File: tests/compile/module_name_arg.srctree [hunk 3]
```diff
@@ -0,0 +1,52 @@
+# Test that we can set module name with --module-name arg to cython
+CYTHON a.pyx
+CYTHON --module-name w b.pyx
+CYTHON --module-name my_module.submod.x c.pyx
+PYTHON setup.py build_ext --inplace
+PYTHON -c "import checks"
```
Comment (scoder) [inline tests/compile/module_name_arg.srctree:6]: If you want to run a script, then run the script.
```suggestion
PYTHON checks.py
```

File: tests/compile/module_name_arg.srctree [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    ('b', False),
+    ('w', True),
+    ('my_module.submod.x', True),
+    ('c', False),
+    ):
+    try:
+        import_module(module_name)
+    except exc:
+        if should_import:
+            assert False, "Connot import module " + module_name
```
Comment (scoder) [inline tests/compile/module_name_arg.srctree:28]:
```suggestion
            assert False, "Cannot import module " + module_name
```

File: Cython/Compiler/Tests/TestCmdLine.py (hunk 1 above)
Comment (matthew-brett) [reply Cython/Compiler/Tests/TestCmdLine.py:508]: Ah - sorry - that was a holdover as I worked out what was being tested.   Here I was confused, because the`-1` flag should generate an error, but there should also be an error for the missing input file name.

Will have a go at the regex check to clarify.

File: Cython/Compiler/Tests/TestCmdLine.py (hunk 2 above)
Comment (matthew-brett) [reply Cython/Compiler/Tests/TestCmdLine.py:536]: I will have a go.

**Review by matthew-brett — COMMENTED** [review]
Thanks for the review.  Will extend a bit with regex for error messages.

**Comment by matthew-brett** [comment]
I think the failure is unrelated.

**Comment by h-vetinari** [comment]
@matthew-brett 
Any update on this? Looking at the diff, there's still 2 unresolved comments at least.

File: Cython/Compiler/Tests/TestCmdLine.py (hunk 1 above)
Comment (matthew-brett) [reply Cython/Compiler/Tests/TestCmdLine.py:508]: I added back the check, with a regex.

**Comment by matthew-brett** [comment]
I've marked the comment that remains as resolved - is there anything else I can do here?

**Comment by rgommers** [comment]
Another gentle ping on this. It's actually important for SciPy's development workflow. Without it, I noticed that there's a ton of unnecessary rebuilding going on. For example, since all extensions built from Cython code depend on the main `__init__.py` without the feature added in this PR, they all get rebuilt as soon as you touch `__init__.py` at all in combination with Meson or any other out-of-place build setup.

Aside from that, it will make moving away from `setuptools` much easier for other projects, because of how much it simplifies working with Cython in Meson. The start of Python 3.12 development is not that far off anymore, and at that point projects are going to want to move away (NumPy and scikit-learn at least are about to start with that).

This is a very small change, now well-tested, that should be quite safe to merge. @scoder any chance this can be merged soon and backported to the 0.29.x branch?

**Comment by h-vetinari** [comment]
> The start of Python 3.12 development is not that far off anymore

Nominally, 3.12 development has kicked off already with the branching of 3.11 off from main.

(+1 to merge & backport)

**Comment by da-woods** [comment]
> Just another ping - is there anything else I can do here?

FWIW this looks good to me. It's just a little larger than I'm happy merging myself.

It might be worth investigating whether it backports cleanly (just so you're prepared to fix any issues if it doesn't) but that's more "advanced preparation" than something to do here.

**Comment by da-woods** [comment]
@scoder this seems like it should probably be in the next release if possible?

**Comment by scoder** [comment]
I think a backport to 0.29.x is reasonable but needs extra work due to various changes in the master branch.

**Comment by h-vetinari** [comment]
> I think a backport to 0.29.x is reasonable but needs extra work due to various changes in the master branch.

I will try a backport later today, unless there is opposition to that or @matthew-brett gets to it before me. :)

**Comment by matthew-brett** [comment]
@h-vetinari - thanks - that would be very helpful.   Let me know if you're stuck for time, I'll try and get to it in a bit.   I'm afraid I've got various deadlines coming up.
